File: analytics/src/main/java/io/appmetrica/analytics/impl/AppMetricaCore.java
package io.appmetrica.analytics.impl;

import android.content.Context;
import android.os.Handler;
import androidx.annotation.AnyThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.AppMetricaConfig;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import io.appmetrica.analytics.coreapi.internal.executors.IHandlerExecutor;
import io.appmetrica.analytics.coreutils.internal.WrapUtils;
import io.appmetrica.analytics.impl.clientcomponents.ClientComponentsInitializerProvider;
import io.appmetrica.analytics.impl.utils.PublicLogger;
import io.appmetrica.analytics.impl.utils.executors.ClientExecutorProvider;

@AnyThread
public class AppMetricaCore implements IAppMetricaCore {

    @NonNull
    private final Context mContext;
    @NonNull
    private final Handler mMetricaHandler;
    @NonNull
    private final ClientTimeTracker mClientTimeTracker;
    @NonNull
    private final ICommonExecutor mDefaultExecutor;
    @NonNull
    private final AppOpenWatcher appOpenWatcher;
    @Nullable
    private volatile AppMetricaUncaughtExceptionHandler mUncaughtExceptionHandler;
    private boolean activated = false;

    AppMetricaCore(@NonNull Context context, @NonNull ClientExecutorProvider clientExecutorProvider) {
        this(
                context.getApplicationContext(),
                clientExecutorProvider.getDefaultExecutor(),
                clientExecutorProvider.getApiProxyExecutor()
        );
    }

    private AppMetricaCore(@NonNull Context context,
                              @NonNull IHandlerExecutor defaultExecutor,
                              @NonNull ICommonExecutor apiProxyExecutor) {
        this(
                context,
                defaultExecutor,
                new ClientTimeTracker(),
                new AppOpenWatcher(apiProxyExecutor)
        );
    }

    @VisibleForTesting
    AppMetricaCore(@NonNull Context context,
                      @NonNull IHandlerExecutor defaultExecutor,
                      @NonNull ClientTimeTracker clientTimeTracker,
                      @NonNull AppOpenWatcher appOpenWatcher) {
        mContext = context;
        mDefaultExecutor = defaultExecutor;
        this.appOpenWatcher = appOpenWatcher;
        PublicLogger.init(mContext);
        defaultExecutor.execute(new Runnable() {
            @Override
            public void run() {
                SdkUtils.logSdkInfo();
            }
        });
        mMetricaHandler = defaultExecutor.getHandler();
        mClientTimeTracker = clientTimeTracker;
        mClientTimeTracker.trackCoreCreation();
        new ClientComponentsInitializerProvider()
            .getClientComponentsInitializer()
            .onCreate();
    }

    @Override
    public synchronized void activate(@NonNull final AppMetricaConfig config,
                                      @NonNull IReporterFactoryProvider reporterFactoryProvider) {
        if (!activated) {
            final boolean crashReportingEnabled = WrapUtils.getOrDefault(
                    config.crashReporting, DefaultValuesForCrashReporting.DEFAULT_REPORTS_CRASHES_ENABLED);
            if (crashReportingEnabled && mUncaughtExceptionHandler == null) {
                mUncaughtExceptionHandler = createUncaughtExceptionHandler(config, reporterFactoryProvider);
                Thread.setDefaultUncaughtExceptionHandler(mUncaughtExceptionHandler);
            }
            final boolean appOpenTrackingEnabled = WrapUtils.getOrDefault(
                    config.appOpenTrackingEnabled,
                    DefaultValues.DEFAULT_APP_OPEN_TRACKING_ENABLED
            );
            if (appOpenTrackingEnabled) {
                appOpenWatcher.startWatching();
            }
            activated = true;
        }
    }

    @Override
    @NonNull
    public Handler getMetricaHandler() {
        return mMetricaHandler;
    }

    @Override
    @NonNull
    public ClientTimeTracker getClientTimeTracker() {
        return mClientTimeTracker;
    }

    @Override
    @NonNull
    public ICommonExecutor getExecutor() {
        return mDefaultExecutor;
    }

    @AnyThread
    @NonNull
    private AppMetricaUncaughtExceptionHandler createUncaughtExceptionHandler(
            @NonNull AppMetricaConfig config,
            @NonNull IReporterFactoryProvider reporterFactoryProvider
    ) {
        return new AppMetricaUncaughtExceptionHandler(
                Thread.getDefaultUncaughtExceptionHandler(),
                ClientServiceLocator.getInstance()
                        .getCrashProcessorFactory()
                        .createCrashProcessors(
                                mContext,
                                config,
                                reporterFactoryProvider
                        )
        );
    }

    @Override
    @NonNull
    public AppOpenWatcher getAppOpenWatcher() {
        return appOpenWatcher;
    }

    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    @Nullable
    AppMetricaUncaughtExceptionHandler getUncaughtExceptionHandler() {
        return mUncaughtExceptionHandler;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/AppMetricaCore.kt
package io.appmetrica.analytics.impl

import android.content.Context
import android.os.Handler
import androidx.annotation.AnyThread
import io.appmetrica.analytics.AppMetricaConfig
import io.appmetrica.analytics.coreapi.internal.executors.IHandlerExecutor
import io.appmetrica.analytics.impl.clientcomponents.ClientComponentsInitializerProvider
import io.appmetrica.analytics.impl.crash.jvm.client.JvmCrashClientController
import io.appmetrica.analytics.impl.utils.executors.ClientExecutorProvider
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger
import io.appmetrica.analytics.logger.common.BaseReleaseLogger

@AnyThread
internal class AppMetricaCore(
    private val context: Context,
    executorProvider: ClientExecutorProvider
) : IAppMetricaCore {

    private val tag = "[AppMetricaCore]"

    override val defaultExecutor: IHandlerExecutor = executorProvider.defaultExecutor
    override val clientTimeTracker = ClientTimeTracker()
    override val defaultHandler: Handler = defaultExecutor.handler
    override val appOpenWatcher = AppOpenWatcher()
    override val jvmCrashClientController = JvmCrashClientController()

    init {
        BaseReleaseLogger.init(context)
        defaultExecutor.execute { SdkUtils.logSdkInfo() }
        clientTimeTracker.trackCoreCreation()
        ClientComponentsInitializerProvider()
            .getClientComponentsInitializer()
            .onCreate()
    }

    @Synchronized
    override fun activate(
        config: AppMetricaConfig?,
        reporterFactoryProvider: IReporterFactoryProvider
    ) {
        DebugLogger.info(tag, "activate with config: ${config?.toJson()?.toString()}")
        if (!AppMetricaFacade.isFullyInitialized()) {
            if (config.shouldCollectCrashes()) {
                jvmCrashClientController.setUpCrashHandler()
                jvmCrashClientController.registerTechnicalCrashConsumers(context, reporterFactoryProvider)
                config?.let {
                    jvmCrashClientController.registerApplicationCrashConsumer(context, reporterFactoryProvider, it)
                }
            } else {
                jvmCrashClientController.clearCrashConsumers()
            }
            if (config.shouldTrackAppOpen()) {
                appOpenWatcher.startWatching()
            } else {
                appOpenWatcher.stopWatching()
            }
            // Config is null means anonymous activation
            if (config != null) {
                AppMetricaFacade.markFullyInitialized()
            }
        } else {
            DebugLogger.info(tag, "Full activation already has been completed")
        }
    }

    private fun AppMetricaConfig?.shouldTrackAppOpen(): Boolean = this?.let {
        it.appOpenTrackingEnabled ?: DefaultValues.DEFAULT_APP_OPEN_TRACKING_ENABLED
    } ?: DefaultValues.DEFAULT_APP_OPEN_TRACKING_ENABLED_FOR_ANONYMOUS_ACTIVATION

    private fun AppMetricaConfig?.shouldCollectCrashes(): Boolean = this?.let {
        it.crashReporting ?: DefaultValuesForCrashReporting.DEFAULT_REPORTS_CRASHES_ENABLED
    } ?: DefaultValuesForCrashReporting.DEFAULT_REPORT_CRASHES_ENABLED_FOR_ANONYMOUS_ACTIVATION
}
